pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -964,6 +985,58 @@ func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {
 	_ = c.XDSUpdater.EDSUpdate(c.ClusterID, string(hostname), ep.Namespace, endpoints)
 }
 
+// called when the overall mTLS ready status of endpoints have changed for a service and the cached Service object must be updated
+func (c *Controller) updateSvcMtlsReady(svc *model.Service) {
+	// TODO GregHanson is this being handled properly?
+	c.RLock()
+	instances := c.externalNameSvcInstanceMap[svc.Hostname]
+	c.RUnlock()
+	if instances != nil {
+		svcMTLSReady := true
+		for _, inst := range instances {
+			if !inst.Service.MTLSReady {
+				svcMTLSReady = false
+				break
+			}
+		}
+		c.Lock()
+		serviceToModify := c.servicesMap[svc.Hostname]
+		serviceToModify.MTLSReady = svcMTLSReady
+		c.servicesMap[svc.Hostname] = serviceToModify
+		c.Unlock()
+		return
+	}
+
+	item, exists, err := c.endpoints.informer.GetStore().GetByKey(kube.KeyFunc(svc.Attributes.Name, svc.Attributes.Namespace))
+
+	if err != nil {
+		log.Infof("get endpoint(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return
+	}
+	if !exists {
+		return
+	}
+
+	svcMTLSReady := true
+	ep := item.(*v1.Endpoints)
+	for _, ss := range ep.Subsets {
+		for _, ea := range ss.Addresses {
+			pod := c.pods.getPodByIP(ea.IP)
+			if !kube.PodMTLSReady(pod) {
+				svcMTLSReady = false
+				break
+			}
+		}
+	}
+
+	c.Lock()
+	serviceToModify := c.servicesMap[svc.Hostname]
+	serviceToModify.MTLSReady = svcMTLSReady
```

Comment:
Can this whole function be just setting this in the map? When it's called it seems we already know if we want to set it to true or false so we don't need to iterate over all endpoints?

Reply:
@howardjohn please check latest.  Only time where all endpoint/instances are iterated over is when `Service.MTLSReady=false` and there is a delete event.  This is designed for the scenario where all instances are mtlsReady=true except for one, and that one instance gets deleted.

Reply:
service level mtls autopilot removed since endpoint level should be available in time for 1.4